shell/platform/windows/client_wrapper/flutter_window_controller.cc

```diff
@@ -0,0 +1,407 @@
+  }
+  return DefWindowProc(hwnd, message, wparam, lparam);
+}
+
+void FlutterWindowController::SendOnWindowCreated(
+    FlutterViewId view_id,
+    std::optional<FlutterViewId> parent_view_id) const {
+  if (channel_) {
+    channel_->InvokeMethod(
+        "onWindowCreated",
+        std::make_unique<EncodableValue>(EncodableMap{
+            {EncodableValue("viewId"), EncodableValue(view_id)},
+            {EncodableValue("parentViewId"),
+             parent_view_id ? EncodableValue(parent_view_id.value())
+                            : EncodableValue()}}));
+  }
+}
+
+void FlutterWindowController::SendOnWindowDestroyed(
+    FlutterViewId view_id) const {
+  if (channel_) {
+    channel_->InvokeMethod(
+        "onWindowDestroyed",
+        std::make_unique<EncodableValue>(EncodableMap{
+            {EncodableValue("viewId"), EncodableValue(view_id)},
+        }));
+  }
+}
+
+void FlutterWindowController::SendOnWindowChanged(FlutterViewId view_id) const {
+  if (channel_) {
+    auto const size{GetWindowSize(view_id)};
+    channel_->InvokeMethod(
+        "onWindowChanged",
+        std::make_unique<EncodableValue>(EncodableMap{
+            {EncodableValue("viewId"), EncodableValue(view_id)},
+            {EncodableValue("size"),
+             EncodableValue(EncodableList{EncodableValue(size.width),
+                                          EncodableValue(size.height)})},
+            {EncodableValue("relativePosition"), EncodableValue()},  // TODO
+            {EncodableValue("isMoving"), EncodableValue()}}));       // TODO
+  }
+}
+
+void FlutterWindowController::HandleCreateWindow(WindowArchetype archetype,
+                                                 MethodCall<> const& call,
+                                                 MethodResult<>& result) {
+  auto const* const arguments{call.arguments()};
+  auto const* const map{std::get_if<EncodableMap>(arguments)};
+  if (!map) {
+    result.Error(kErrorCodeInvalidValue, "Method call argument is not a map.");
+    return;
+  }
+
+  std::wstring const title{ArchetypeToWideString(archetype)};
+
+  auto const size_list{
+      GetListValuesForKeyOrSendError<int, 2>("size", map, result)};
+  if (!size_list) {
+    return;
+  }
+  if (size_list->at(0) < 0 || size_list->at(1) < 0) {
+    result.Error(kErrorCodeInvalidValue,
+                 "Values for 'size' key (" + std::to_string(size_list->at(0)) +
+                     ", " + std::to_string(size_list->at(1)) +
+                     ") must be nonnegative.");
+    return;
+  }
+
+  if (auto const data_opt{CreateFlutterWindow(
+          title, {.width = size_list->at(0), .height = size_list->at(1)},
+          archetype)}) {
+    auto const& data{data_opt.value()};
+    result.Success(EncodableValue(EncodableMap{
+        {EncodableValue("viewId"), EncodableValue(data.view_id)},
+        {EncodableValue("archetype"),
+         EncodableValue(static_cast<int>(data.archetype))},
+        {EncodableValue("size"),
```

Comment:
Might be easier to send "width" and "height" - less marshalling required.

Reply:
The current approach aims to maintain symmetry between the Dart API and the serialized structure, although we haven't given it much thought yet. I agree that a flat structure would simplify the serialization, but the nested approach may be more readable and intuitive as we add more attributes.

Maybe we should decide whether to adopt a flat structure (easier to serialize, but possibly cluttering the payload with many keys) or a nested structure (mimicking the Dart syntax but a bit more complex to parse) and write this down in the design doc.

Nested structure (using `createPopup` as an example):
```json
{
  "parentViewId": 42,
  "preferredSize": { "width": 800, "height": 600 },
  "anchorRect": { "left": 0, "top": 0, "width": 100, "height": 200 },
  "positionerParentAnchor": "topLeft",
  "positionerChildAnchor": "center",
  "positionerOffset": { "dx": 10, "dy": 20 },
  "positionerConstraintAdjustment": ["slideX", "slideY"]
}
```

Flat structure:
```json
{
  "parentViewId": 42,
  "preferredWidth": 800,
  "preferredHeight": 600,
  "anchorRectLeft": 0,
  "anchorRectTop": 0,
  "anchorRectWidth": 100,
  "anchorRectHeight": 200,
  "positionerParentAnchor": "topLeft",
  "positionerChildAnchor": "topLeft",
  "positionerOffsetX": 10,
  "positionerOffsetY": 20,
  "positionerConstraintAdjustment": ["slideX", "slideY"]
}
```